_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+# Spike: Resolve included models
+
+Consider the following domain model(s):
+
+- A model called `Category` with properties `id`, `name`.
+- A model called `Product` with properties `id`, `name`, `categoryId`
+- A `hasMany` relation (Category has many products)
+
+Now consider the following code to retrieve all categories with all related
+products (perhaps to render a home page of a product catalog):
+
+```ts
+categoryRepo.find({include: [{relation: 'products'}]});
+```
+
+## The problem
+
+How to fetch products for each category found?
+
+Additional constraints:
+
+- The target model (`Product`) can be backed by a different database than the
+  source model (`Category`). For example, we can use MongoDB to store categories
+  and MySQL to store products.
+
+- We need to test relations against real databases to ensure we are correctly
+  handling database-specific quirks like:
+  - a limit on the number of items that can be included in `inq`
+  - coercion between `ObjectID` vs. `string` for MongoDB.
+
+## Proposed solution
+
+### 1. Introduce the concept of `InclusionResolver`
+
+An inclusion resolver is a class that can fetch target models for the given list
+of source model instances. The idea is to create a different inclusion resolver
+for each relation type, e.g. `HasManyInclusionResolver`, etc.
+
+```ts
+export interface InclusionResolver {
```

Comment:
Can it be simplified as a function?

Reply:
Most likely yes, a resolver can be a simple function too. I'll take a look.

However, if we want `HasManyRepositoryFactory` to provide inclusion resolver too (as discussed in https://github.com/strongloop/loopback-next/pull/3387#discussion_r305025912), then we need to store the resolver function as a property on that factory function.

Let me think about this some more.

Reply:
Done in a7260ff